src/game_api/mod_api.cpp

```diff
@@ -0,0 +1,455 @@
+#include "search.hpp"
+#include "spawn_api.hpp"
+#include "state.hpp"
+#include "steam_api.hpp"
+#include "strings.hpp"
+#include "virtual_table.hpp" // for VTABLE_OFFSET
+
+static bool g_forward_blocked_events{false};
+
+static int64_t global_update_count{0};
+static int64_t global_frame_count{0};
+
+int64_t get_global_frame_count()
+{
+    return global_frame_count;
+}
+int64_t get_global_update_count()
+{
+    return global_update_count;
+}
+
+bool get_forward_events()
+{
+    return g_forward_blocked_events;
+}
+
+static bool g_godmode_player_active = false;
+static bool g_godmode_companions_active = false;
+
+void ModAPI::godmode(bool g)
+{
+    g_godmode_player_active = g;
+}
+
+void ModAPI::godmode_companions(bool g)
+{
+    g_godmode_companions_active = g;
+}
+
+void ModAPI::darkmode(bool g)
+{
+    static const size_t addr_dark = get_address("force_dark_level");
+
+    if (g)
+    {
+        write_mem_recoverable("darkmode", addr_dark, "\x90\x90"sv, true);
+    }
+    else
+    {
+        recover_mem("darkmode");
+    }
+}
+
+bool is_active_player(Entity* e)
+{
+    auto state = State::ptr();
+    for (uint8_t i = 0; i < MAX_PLAYERS; i++)
+    {
+        auto player = state->items->player(i);
+        if (player && player == e)
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+using OnDamageFun = bool(Entity*, Entity*, int8_t, uint32_t, float*, uint8_t, uint16_t, uint8_t, bool);
+OnDamageFun* g_on_damage_trampoline{nullptr};
+bool on_damage(Entity* victim, Entity* damage_dealer, int8_t damage_amount, uint32_t unknown1, float* velocities, uint8_t unknown2, uint16_t stun_amount, uint8_t iframes, bool unknown3)
+{
+    if (g_godmode_player_active && is_active_player(victim))
+    {
+        return false;
+    }
+    if (g_godmode_companions_active && !is_active_player(victim) && (victim->type->search_flags & 1) == 1)
+    {
+        return false;
+    }
+
+    return g_on_damage_trampoline(victim, damage_dealer, damage_amount, unknown1, velocities, unknown2, stun_amount, iframes, unknown3);
+}
+
+using OnInstaGibFun = void(Entity*, bool, size_t);
+OnInstaGibFun* g_on_instagib_trampoline{nullptr};
+void on_instagib(Entity* victim, bool destroy_corpse, size_t param_3)
+{
+    if (g_godmode_player_active && is_active_player(victim))
+    {
+        return;
+    }
+    if (g_godmode_companions_active && !is_active_player(victim) && (victim->type->search_flags & 1) == 1)
+    {
+        return;
+    }
+
+    const bool skip_orig = pre_entity_instagib(victim) && !(victim->as<Movable>()->health == 0);
+
+    if (!skip_orig)
+    {
+        g_on_instagib_trampoline(victim, destroy_corpse, param_3);
+    }
+}
+
+void hook_godmode_functions()
+{
+    static bool functions_hooked = false;
+    if (!functions_hooked)
+    {
+        auto memory = Memory::get();
+        auto addr_damage = memory.at_exe(get_virtual_function_address(VTABLE_OFFSET::CHAR_ANA_SPELUNKY, 48));
```

Comment:
I know you didn't write this code, but wondering why this is done via Detour probably better to do it via `hook_vtable` ?